Landing detailed

swforshort opened the pull request:
It safely lands the drone, if and only if it is in hover.

(I figured if I messed up it might as well be a fork. Especially when I dropped the file in the wrong spot initially.)

swforshort committed:
Add files via upload

changed landing to slowly drop to a standstill at 0, but will not do so if in forward flight mode.

swforshort committed:
Add files via upload

swforshort committed:
Delete TrifanHarness.cpp

swforshort commented:
The '3' commits is really just one, since two were adding and deleting the file from the wrong spot.

evizitei commented on src/TrifanHarness.cpp:
```diff
@@ -76,9 +76,21 @@ int main()
+                while(gps->getAltitude() >0.0)
+                    ctrl->updateMotors(2900);
+                ctrl->updateMotors(3000);
+            }
```
yeah, this works; the kind of downside here is the tight loop that keeps updating the motors, which takes the mutex on the flight controller each time unnecessarily and it will do that hundreds of times per second, so that seems wasteful. Once your RPMs have been set, there's no need to keep resetting them; maybe instead you should set the motors to 2500 BEFORE the loop, and then check in once a second or something by using the sleep command to wait (search the project for `std::this_thread::sleep_for` for an example, the FlightLog and Simulator classes both do something like this).  This would even let you keep printing feedback to the console.  Like you could say on each loop, say " .... [current altitude] ...\n" . to keep the user aware and then sleep for another second. Then when each loop condition is broken readjust the motors up to the next RPM level ONE time and go into another sleep/print loop.  Does that make sense?

swforshort replied:
what about this?

swforshort committed:
fixed updating motors continually

added in timers and such to measure and adjust motors only once per tier

evizitei commented on src/TrifanHarness.cpp:
```diff
@@ -76,16 +77,39 @@ int main()
+                    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+                }
+
+                std::cout << "Landed. \n";
```
No, I think this is way too risky.  This will *technically* work in this specific simulator because you can kind of estimate how much time it's going to take, but that would bind it just to this simulator; and really, as soon as I add any random noise to the simulator (wind gusts, variations in air pressure, etc), this is totally broken because it's not checking it's altitude on the way down.  I think the prior strategy we talked about is the way to go. Set your descent rotor speed based on altitude, then in a "while(true)" loop sleep for 1 second and check altitude and sleep for a second and check altitude etc until you get below the next threshold, then adjust speed and go into another loop, continue until landed

swforshort committed:
no more timing

I changed out the timed switches to the checking once a second. Everything works as intended still, except two issues:

1) The altitude std::cout statement isn't writing inside the loop. It's not throwing any errors, it just isn't writing it anywhere I could see.

2) Currently I have the values hard coded in, but, if we switch to a more realistic model, it will need to change.  I added in the values and comments to explain that.  The gist of it is that I would need to make it a portion of the motor speed needed to have it hold stable altitude.
